ext/fg/js/frontend-api-sender.js

```diff
@@ -0,0 +1,128 @@
+        this.callbacks = {};
+        this.disconnected = false;
+
+        this.port = chrome.runtime.connect(null, {name: 'backend-api-forwarder'});
+        this.port.onDisconnect.addListener(this.onDisconnect.bind(this));
+        this.port.onMessage.addListener(this.onMessage.bind(this));
+    }
+
+    invoke(action, params, target) {
+        if (this.disconnected) {
+            return Promise.reject('Disconnected');
+        }
+
+        const id = FrontendApiSender.generateId(16);
+        if (this.callbacks.hasOwnProperty(id)) {
+            return Promise.reject('Invalid ID');
+        }
+
+        return new Promise((resolve, reject) => {
+            const info = {id, resolve, reject, ack: false, timer: null};
+            this.callbacks[id] = info;
+            info.timer = setTimeout(() => this.onError(id, 'Timeout (ack)'), this.ackTimeout);
+
+            this.port.postMessage({id, action, params, target, senderId: this.senderId});
+        });
+    }
+
+    onMessage({type, id, data, senderId}) {
+        if (senderId !== this.senderId) { return; }
+        switch (type) {
+            case 'ack':
+                this.onAck(id);
+                break;
+            case 'result':
+                this.onResult(id, data);
+                break;
+        }
+    }
+
+    onDisconnect() {
+        this.disconnected = true;
+
+        const ids = Object.keys(this.callbacks);
+        for (const id of ids) {
+            this.onError(id, 'Disconnected');
+        }
+    }
+
+    onAck(id) {
+        if (!this.callbacks.hasOwnProperty(id)) {
+            console.warn(`ID ${id} not found for ack`);
+            return;
+        }
+
+        const info = this.callbacks[id];
+        if (info.ack) {
+            console.warn(`Request ${id} already ack'd`);
+            return;
+        }
+
+        info.ack = true;
+        clearTimeout(info.timer);
+        info.timer = setTimeout(() => this.onError(id, 'Timeout (response)'), this.responseTimeout);
+    }
+
+    onResult(id, data) {
+        if (!this.callbacks.hasOwnProperty(id)) {
+            console.warn(`ID ${id} not found`);
+            return;
+        }
+
+        const info = this.callbacks[id];
+        if (!info.ack) {
+            console.warn(`Request ${id} not ack'd`);
+            return;
+        }
+
+        delete this.callbacks[id];
+        clearTimeout(info.timer);
+        info.timer = null;
+
+        if (typeof data.error === 'string') {
+            info.reject(data.error);
+        } else {
+            info.resolve(data.result);
+        }
+    }
+
+    onError(id, reason) {
+        if (!this.callbacks.hasOwnProperty(id)) { return; }
+        const info = this.callbacks[id];
+        delete this.callbacks[id];
+        info.timer = null;
+        info.reject(reason);
+    }
+
+    static generateId(length) {
+        let id = '';
+        for (let i = 0; i < length; ++i) {
+            id += Math.floor(Math.random() * 256).toString(16).padStart(2, '0');
```

Comment:
This really feels hazardous, but if you can't connect to the backend any other way, it should be fine if the random string is long enough. Do you think this could be sequential instead or would it cause the responses to be delivered to wrong popups?

Reply:
There are 3 places a pseudorandom 32-character ID is used:
1. [PopupProxyHost L#53](https://github.com/FooSoft/yomichan/pull/185/files/b64b2df1cc0de3af4629df95a55647443e1e71c1#diff-0c0192ac230f017dd32f41afd4194c98R53) in ```generateUniqueId```
  This could (and likely should) be replaced with a counter which is incremented for each popup created.
2. [FrontendApiSender L#38](https://github.com/FooSoft/yomichan/pull/185/files/b64b2df1cc0de3af4629df95a55647443e1e71c1#diff-4ee0bdacc6e5034db6a70faadee030a9R38) in ```invoke```
  This could also be replaced with a counter which is incremented for each message sent. This ID is used to find the reference to the callback.
3. [FrontendApiSender L#22](https://github.com/FooSoft/yomichan/pull/185/files/b64b2df1cc0de3af4629df95a55647443e1e71c1#diff-4ee0bdacc6e5034db6a70faadee030a9R22) in ```constructor```
  This is the only unique identifier which is truly necessary. I believe this was the unique ID which was used to determine the correct destination when receiving callback messages when there are iframes on the page. The [```Port```](https://developer.chrome.com/extensions/runtime#type-Port) type can have multiple receivers on either end, so this is why this was necessary.

The uniqueness of the psuedorandom IDs should be adequate for the 3rd case, as there only needs to be enough unique IDs as the amount of frames on the page. Something like [```Crypto.getRandomValues```](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) could be used instead of ```Math.random```, but this wouldn't affect anything in practice. The intent was to generate GUID-like strings using random input, which seems to be a valid way of generating GUIDs (https://www.ietf.org/rfc/rfc4122.txt section 4.4).

Reply:
Updated to use incrementing index for IDs for 1 and 2.